src/Microsoft.Azure.Batch.SoftwareEntitlement.Server/Controllers/SoftwareEntitlementsController.cs

```diff
@@ -104,6 +88,73 @@ public SoftwareEntitlementsController(ServerOptions serverOptions, ILogger logge
             }
         }
 
+        private class EntitlementRequestParameters
+        {
+            public string Token { get; set; }
+
+            public string ApplicationId { get; set; }
+
+            public IPAddress RemoteAddress { get; set; }
+        }
+
+        /// <summary>
+        /// Attempts to extracts all the parameters required for validating an entitlement request.
+        /// Any error here reflects a badly formed request.
+        /// </summary>
+        /// <param name="apiVersion">The API version from the query string</param>
+        /// <param name="requestBody">The information in the request body</param>
+        /// <returns>
+        /// A tuple in which either the <see cref="EntitlementRequestParameters"/> and token values
+        /// are present if the request was well formed, or an informative error otherwise.
+        /// </returns>
+        private (EntitlementRequestParameters Parameters, string Error) TryExtractParameters(
+            string apiVersion,
+            SoftwareEntitlementRequestBody requestBody)
+        {
```

Comment:
Much of this method is concerned with checking `SoftwareEntitlementRequestBody` - is this a case of the *feature envy* design smell? If so, perhaps moving the validation onto the body would make sense.

Reply:
Agree that it might make sense to move some of this logic somewhere else, particularly as more API versions are added. I think that might make sense as a separate PR, as we probably have a few different options (not sure about adding anything to `SoftwareEntitlementRequestBody` though; shouldn't that just be a POCO for deserializing the HTTP request?). For now, I'm just trying to gather the logic into one place to make future refactoring a bit easier.

Reply:
Keeping `SoftwareEntitlementRequestBody` as a POCO doesn't mean that methods are forbidden - I see validation & transformation methods as very much having a place on simple data-transfer objects like this.

> For now, I'm just trying to gather the logic into one place to make future refactoring a bit easier.

Sure, no problem.

Reply:
I'm not opposed to having a few simple, consistent methods on the `Body` class. But as the same class is used for all the different API versions I think it could get out of control.

If it does get much more complicated, I could see us using something like a strategy pattern to build the appropriate request based on the supplied API version. On the other hand, if it's simple enough to have all the logic in one class I don't have a problem with keeping it all in the `Controller` so you don't have to go scooting around several files to work out what's going on.

Reply:
I believe that ASP.NET routing would allow us to route to a different controller method for each `apiVersion`, so we'd be able to use a different class to represent each kind of request.

Reply:
Good idea. Might give that a try after my other changes, otherwise it'll take forever to get those done...

Reply:
Agreed, we don't need worry about it right now with only two versions, but something to keep in mind when we get to v3.